### PR TITLE
[Main Course] - TensorFlow Recommender 2

### PR DESCRIPTION
**Pull request by Tina0330**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

# How Has This Been Tested?

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: docs/Main_Course/Deep_Learning/2021-12-27-TensorFlow-Recommender2.md [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+随机打乱数据，并取80%为训练数据集，其余的20%为测试数据集：
+```Python
+tf.random.set_seed(42)
+shuffled = ratings.shuffle(100_000, seed=42, reshuffle_each_iteration=False)
+
+train = shuffled.take(80_000)
+test = shuffled.skip(80_000).take(20_000)
+```
+
+为原分类型数据创造嵌入的数组，这里使用的是用连续整数来匹配电影标题和用户ID的每行数据：
```
Comment (jhqian0909) [inline docs/Main_Course/Deep_Learning/2021-12-27-TensorFlow-Recommender2.md:60]: 这句话可能会让人看不明白

File: docs/Main_Course/Deep_Learning/2021-12-27-TensorFlow-Recommender2.md [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+movie_titles = ratings.batch(1_000_000).map(lambda x: x["movie_title"])
+user_ids = ratings.batch(1_000_000).map(lambda x: x["user_id"])
+
+unique_movie_titles = np.unique(np.concatenate(list(movie_titles)))
+unique_user_ids = np.unique(np.concatenate(list(user_ids)))
+```
+
+### 2.2 搭建排序模型
+
+先设置用户和电影的`Embedding`，设置模型层和预测层，接着用`call()`函数输入input，从而返回ranking的结果：
```
Comment (jhqian0909) [inline docs/Main_Course/Deep_Learning/2021-12-27-TensorFlow-Recommender2.md:72]: 可以简单介绍一下

File: docs/Main_Course/Deep_Learning/2021-12-27-TensorFlow-Recommender2.md [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+```
+由下图所示，用未经过训练的模型为用户号为9推荐电影《One Flew Over the Cuckoo's Nest (1975)》，预测的可能性为0.016：
+
+<figure>
+  <img src="https://cdn.jsdelivr.net/gh/BulletTech2021/Pics/img/rank_result11.png" width="500"/>
+</figure>
+
+
+为了让模型有更好的表现，在原有的模型任务中加入`MSE`均方误差的损失函数和`RMSE`均方根误差的评估函数：
```
Comment (jhqian0909) [inline docs/Main_Course/Deep_Learning/2021-12-27-TensorFlow-Recommender2.md:124]: 应该不能说让它有更好的表现吧 毕竟前面都没开始训练

File: docs/Main_Course/Deep_Learning/2021-12-27-TensorFlow-Recommender2.md [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+import tensorflow as tf
+## TensorFlow Dataset Resource
+import tensorflow_datasets as tfds
+
+## TensorFlow 推荐系统
+import tensorflow_recommenders as tfrs
+
+```
+
+从`TensorFlow Dataset`中引入和信息检索模型相同的`movielens`电影数据集，并只保留以下三个变量：
```
Comment (EchoTang04) [inline docs/Main_Course/Deep_Learning/2021-12-27-TensorFlow-Recommender2.md:40]: 前言里说排序模型是对第一步检索模型的结果再做调整，排序模型的input好像没有体现这一点？感觉还是用的最开始的input吗？

**Review by EchoTang04 — APPROVED** [review]
LGTM, 感谢分享~

**Review by Realvincentyuan — APPROVED** [review]

File: docs/Main_Course/Deep_Learning/2021-12-27-TensorFlow-Recommender2.md (hunk 3 above)
Comment (Tina0330) [reply docs/Main_Course/Deep_Learning/2021-12-27-TensorFlow-Recommender2.md]: 已修改

**Pull request merged**